libsolidity/ast/Types.cpp

```diff
@@ -172,7 +172,8 @@ void StorageOffsets::computeOffsets(TypePointers const& _types)
 			++slotOffset;
 			byteOffset = 0;
 		}
-		solAssert(slotOffset < bigint(1) << 256 ,"Object too large for storage.");
+		if (slotOffset >= bigint(1) << 256)
+			BOOST_THROW_EXCEPTION(Error(3666_error, Error::Type::TypeError) << util::errinfo_comment("Object too large for storage."));
```

Comment:
I am not completely sure if it is better to re-use `Error` here or to create a new exception class.

Now `Error` has a dual role, Almost always it is simply an item in `ErrorReporter::m_errorList`. And there are just a couple of cases where it gets thrown as an exception.

I think such dual role would be fine if we throw `Error` more often. Otherwise we may prefer to treat it as a regular data struct, not as an exception.

Reply:
Is it possible to add a test in the TypeChecker to see if the constructed array is too big?

Reply:
I can add to `bool TypeChecker::visit(ContractDefinition const&)`:
```
try
{
    ContractType(_contract).stateVariables();
}
catch (Error const& _e)
{
    m_errorReporter.error(_e.errorId(), _e.type(), _contract.location(), _e.what());
}
```

Would it be the right way? (I do not quite like that I'm introducing a code pattern which was not used before.)

Reply:
I think it should go in `bool TypeChecker::visit(VariableDeclaration const& _variable)`
Maybe this computation can be done there locally?

Reply:
Maybe we can check size of individual arrays in `bool TypeChecker::visit(VariableDeclaration const&)`, but I am not sure about the total size. As we are summing up storage sizes for different types (see `StorageOffsets::computeOffsets()`), it might be hard to do it locally.

Reply:
@hrkrshnn Ahh, I did not see that your comment was edited in the meantime... So my answer above is to its first edition :).